### ISSUE TITLE
Tunnelfront connection failures causing tiller issues

### ISSUE THREAD

**Issue by jlogsdon** [Azure/AKS#1311]
**What happened**: In the East US region my helm chart installation is consistently stalling while waiting for deployments to come up. Five times in a row I have seen the tiller pod stop output at a `2019/11/13 16:10:13 Pod is not ready` state. I've found that shortly after the last log is output in tiller I see this in tunnelfront:

```
Timeout, server ENDPOINT.hcp.eastus.azmk8s.io not responding.
[ Wed Nov 13 16:10:37 UTC 2019 ] INF: Tunnel front and end both server.version == local.version, no rotation needed
[ Wed Nov 13 16:10:37 UTC 2019 ] INF: Ssh to tunnelend is not connected.. attempting to connect
Warning: Permanently added the RSA host key for IP address '[xx.xx.xx.xx]:9000' to the list of known hosts.
[ Wed Nov 13 16:10:37 UTC 2019 ] INF: connection to tunnelend established
```

Eventually the tiller deploy times out:

```
[tiller] 2019/11/13 16:26:00 warning: Release "NAME" failed: Get https://ENDPOINT.hcp.eastus.azmk8s.io:443/apis/apps/v1/namespaces/default/deployments/DEPLOY: read tcp xx.xx.xx.xx:56660->xx.xx.xx.xx:443: read: connection timed out
```

Purging and restarting the deploy only triggers the issue again. We've tried three times in a row on the same cluster and it always times out.

**What you expected to happen**: tunnelfront should not lose connection and break the active helm deployment.

**How to reproduce it (as minimally and precisely as possible)**: This is with a private helm chart, but any chart that takes more than a few minutes to provision (such as elasticsearch lately, thanks to other performance issues we've noticed) should reproduce the issue.

**Anything else we need to know?**: We have also tested in Central US and the issue does not occur there. This appears to be isolated to East US.

**Environment**:
- Kubernetes version (use `kubectl version`): Tried with `1.13.11`, `1.13.12`, and `1.14.8`. All three consistently show the issue in East US.
- Size of cluster: 3
- General description of workloads in the cluster: elasticsearch, redis, with nodejs and python services
- Others:

**Comment by jlogsdon** [comment]
Can we get any response on this? It's _still_ happening quite frequently for us.

**Comment by jnoller** [comment]
@jlogsdon Most of the tunnel timeouts I've root cause have been triggered by this: https://aka.ms/aks/io-throttle-issue

**Comment by jlogsdon** [comment]
Thank you for the link, I'll do the investigation steps and see if that is it.

**Comment by jlogsdon** [comment]
It's hard to tell if this is the same issue. I've tried with these different node pool setups and still have the tunnel timeout problem:

1. `Standard_D2s_v3` with a 100GiB disk
2. `Standard_D2s_v3` with a 500GiB disk
3. `Standard_D4s_v3` with a 500GiB disk

With the last one I've had issues getting prometheus-operator setup per the instructions in #1373, once I run the sed commands and restart kubelet the `metrics-server` deployment stops functioning:

```
E0131 21:02:12.038281       1 reststorage.go:135] unable to fetch node metrics for node "aks-logrocket-39218709-vmss000001": no metrics known for node
E0131 21:02:12.038313       1 reststorage.go:135] unable to fetch node metrics for node "aks-logrocket-39218709-vmss000002": no metrics known for node
E0131 21:02:12.038319       1 reststorage.go:135] unable to fetch node metrics for node "aks-logrocket-39218709-vmss000000": no metrics known for node
```

Using the Portal metrics, however, with the last setup I see that the maximum write IOPS peaked just over 200 which should be nothing for a D4s_v3 with 500GiB disk.

![image](https://user-images.githubusercontent.com/88460/73574622-25bf9980-4444-11ea-9e10-ff6aab18cde6.png)

**Comment by jnoller** [comment]
You need to raise the size of the disk to at least 1 TB or ideally, 2TB to maximize the likliehood you won't get throttled. A 500gb OS disk has a max IOPS of 2300/s - so your OS disk to your VM is still throttled leading to the failure. If you look in the kublet logs on the nodes you will see errors like this:

```
Jan 30 23:23:40 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:23:40.383506    4382 container_manager_linux.go:457] [ContainerManager]: Discovered runtime cgroups name: /system.slice/docker.service
Jan 30 23:24:31 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:24:31.646552    4382 log.go:172] http: superfluous response.WriteHeader call from k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader (httplog.go:184)
Jan 30 23:25:07 aks-agentpool-57418505-vmss000000 kubelet[4382]: W0130 23:25:07.241486    4382 reflector.go:302] object-"kube-system"/"azure-ip-masq-agent-config": watch of *v1.ConfigMap ended with: too old resource version: 3244649 (3245058)
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.075369    4382 kubelet.go:1888] SyncLoop (ADD, "api"): "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4_default(47fba903-9c81-4c87-ab6c-9794e0829414)"
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.189935    4382 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "tls-proxy-secret" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-tls-proxy-secret") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.189984    4382 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "prop-prometheus-operator-operator-token-2jfrz" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-prop-prometheus-operator-operator-token-2jfrz") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.237446    4382 kubelet.go:1888] SyncLoop (ADD, "api"): "prop-prometheus-node-exporter-q7fvn_default(8861c2f4-e635-43d3-a6af-f826801ed969)"
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.290337    4382 reconciler.go:248] operationExecutor.MountVolume started for volume "tls-proxy-secret" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-tls-proxy-secret") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.290387    4382 reconciler.go:248] operationExecutor.MountVolume started for volume "prop-prometheus-operator-operator-token-2jfrz" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-prop-prometheus-operator-operator-token-2jfrz") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.298654    4382 operation_generator.go:713] MountVolume.SetUp succeeded for volume "tls-proxy-secret" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-tls-proxy-secret") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.304311    4382 operation_generator.go:713] MountVolume.SetUp succeeded for volume "prop-prometheus-operator-operator-token-2jfrz" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-prop-prometheus-operator-operator-token-2jfrz") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.390753    4382 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "sys" (UniqueName: "kubernetes.io/host-path/8861c2f4-e635-43d3-a6af-f826801ed969-sys") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.390803    4382 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "prop-prometheus-node-exporter-token-xkddz" (UniqueName: "kubernetes.io/secret/8861c2f4-e635-43d3-a6af-f826801ed969-prop-prometheus-node-exporter-token-xkddz") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.390899    4382 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "proc" (UniqueName: "kubernetes.io/host-path/8861c2f4-e635-43d3-a6af-f826801ed969-proc") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
```

In the docker logs, similar lifecycle failures: 

```
debug=false pid=8124
Jan 30 23:28:15 aks-agentpool-57418505-vmss000000 dockerd[3349]: time="2020-01-30T23:28:15.202303554Z" level=info msg="shim reaped" id=669347180a9aefac468e71cb845b439b1b7fc1fb13c0a975a59bed56a38f528a
Jan 30 23:28:15 aks-agentpool-57418505-vmss000000 dockerd[3349]: time="2020-01-30T23:28:15.212570837Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Jan 30 23:28:15 aks-agentpool-57418505-vmss000000 dockerd[3349]: time="2020-01-30T23:28:15.212677538Z" level=warning msg="669347180a9aefac468e71cb845b439b1b7fc1fb13c0a975a59bed56a38f528a cleanup: failed to unmount IPC: umount /var/lib/docker/containers/669347180a9aefac468e71cb845b439b
Jan 30 23:28:16 aks-agentpool-57418505-vmss000000 dockerd[3349]: time="2020-01-30T23:28:16.383424549Z" level=info msg="shim reaped" id=fb9452a1c6fae7ea7effabae5266f0a9122e4f51b6eacd9f40f4a0aee1f7b549
Jan 30 23:28:16 aks-agentpool-57418505-vmss000000 dockerd[3349]: time="2020-01-30T23:28:16.393830634Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
Jan 30 23:28:17 aks-agentpool-57418505-vmss000000 dockerd[3349]: time="2020-01-30T23:28:17.503209646Z" level=info msg="shim containerd-shim started" address="/containerd-shim/moby/2dcfc67a6a4774409a7200f91a13dd4fd37bd7d7723cc33207fe9817c981279e/shim.sock" debug=false pid=8412
```

You VM SKU is fine - it's peak IOPS quota is so high that with a single disk, you can never ever maximize the IOPS per second. Except that the OS disk is a NAS (network attached storage) drive with a hard IOPS quota of 2300 vs your SKU limit.

You will see patterns like this if you check the Disk IO Azure insights chart:

![p4-diskbusy-iops](https://user-images.githubusercontent.com/51528/73575158-d02fc080-4434-11ea-8c41-ee1bb5f4cf57.png)

**Comment by jlogsdon** [comment]
I just did a test with a 1TiB OS disk and the issue persists.

>  A 500gb OS disk has a max IOPS of 2300/s - so your OS disk to your VM is still throttled leading to the failure.

This is where I get confused. According to metrics provided _by azure_ the IOPS usage hardly goes over 200! How am I getting quota limited if I'm ~2100 away from the limit?

**Comment by jnoller** [comment]
@jlogsdon The quota that you're throttled on is not the quota of the Virtual Machine, *its the maximum quota of the worker node operating system disks*:

![disk-sizes](https://user-images.githubusercontent.com/51528/73577714-e50f5280-443a-11ea-8c7f-2bb989426e98.png)
![vm-sizes](https://user-images.githubusercontent.com/51528/73577729-ee002400-443a-11ea-93d5-e6bb38b173bf.png)

Take an example DS3v2 - the peak IOPS is 12800 IOPS (io operations/sec) this means that in order to saturate the VM you would need to use at least 4x3200 IOPS capable disks - those disks would need to be a minimum 1TB in size each.

In the case of the OS disk, AKS defaults to 30GB (terraform, old clusters) or 100GB - looking in the disk size chart above, that maps to a P10 disk with a peak IOPS value of 500 

This means with a DS3v_2 has a peak IOPS to the **operating system disk** which is a network attached Azure managed disk with the quota above ^

**Comment by jlogsdon** [comment]
Okay, so taking my most recent test this should hold true:

1. The VM (a `D4s_v3`) has a maximum IOPS of 6400.
2. The OS disk (`1000GiB`) should be a P30, with a maximum IOPS of 5000.

This means the machine is limited to 5000 IOPS, correct? If the tunnelfront connection loss is due to disk quota throttling how are we hitting that with only _200_ IOPS used?

**Comment by jnoller** [comment]
Yes, the maximum would be 5000 - what is the disk busy VMSS view and Disk IO azure insights reports show for your system? I am not saying 100% my issue is your issue, but I've seen (and can recreate this on larger systems).

This is mine, to cause this I ran a `helm install <chart>` - the cluster is otherwise idle:

![p4-diskbusy-iops](https://user-images.githubusercontent.com/51528/73579359-6a950180-443f-11ea-9b24-addc95d7ed8f.png)
![p4-diskbusyspike](https://user-images.githubusercontent.com/51528/73579360-6b2d9800-443f-11ea-9927-62a0475c3745.png)

The throttle may / could be kicking in briefly which causes the docker runtime to fail - maybe not, you would need to log into the AKS worker nodes and look at the logs, and the pod logs. I'm not saying your issue _is_ this issue - that could be an NSG, or something else. However this will rule this out:


```
Jan 30 23:23:40 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:23:40.383506    4382 container_manager_linux.go:457] [ContainerManager]: Discovered runtime cgroups name: /system.slice/docker.service
Jan 30 23:24:31 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:24:31.646552    4382 log.go:172] http: superfluous response.WriteHeader call from k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader (httplog.go:184)
Jan 30 23:25:07 aks-agentpool-57418505-vmss000000 kubelet[4382]: W0130 23:25:07.241486    4382 reflector.go:302] object-"kube-system"/"azure-ip-masq-agent-config": watch of *v1.ConfigMap ended with: too old resource version: 3244649 (3245058)
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.075369    4382 kubelet.go:1888] SyncLoop (ADD, "api"): "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4_default(47fba903-9c81-4c87-ab6c-9794e0829414)"
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.189935    4382 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "tls-proxy-secret" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-tls-proxy-secret") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.189984    4382 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "prop-prometheus-operator-operator-token-2jfrz" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-prop-prometheus-operator-operator-token-2jfrz") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.237446    4382 kubelet.go:1888] SyncLoop (ADD, "api"): "prop-prometheus-node-exporter-q7fvn_default(8861c2f4-e635-43d3-a6af-f826801ed969)"
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.290337    4382 reconciler.go:248] operationExecutor.MountVolume started for volume "tls-proxy-secret" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-tls-proxy-secret") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.290387    4382 reconciler.go:248] operationExecutor.MountVolume started for volume "prop-prometheus-operator-operator-token-2jfrz" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-prop-prometheus-operator-operator-token-2jfrz") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.298654    4382 operation_generator.go:713] MountVolume.SetUp succeeded for volume "tls-proxy-secret" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-tls-proxy-secret") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.304311    4382 operation_generator.go:713] MountVolume.SetUp succeeded for volume "prop-prometheus-operator-operator-token-2jfrz" (UniqueName: "kubernetes.io/secret/47fba903-9c81-4c87-ab6c-9794e0829414-prop-prometheus-operator-operator-token-2jfrz") pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4" (UID: "47fba903-9c81-4c87-ab6c-9794e0829414")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.390753    4382 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "sys" (UniqueName: "kubernetes.io/host-path/8861c2f4-e635-43d3-a6af-f826801ed969-sys") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.390803    4382 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "prop-prometheus-node-exporter-token-xkddz" (UniqueName: "kubernetes.io/secret/8861c2f4-e635-43d3-a6af-f826801ed969-prop-prometheus-node-exporter-token-xkddz") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.390899    4382 reconciler.go:203] operationExecutor.VerifyControllerAttachedVolume started for volume "proc" (UniqueName: "kubernetes.io/host-path/8861c2f4-e635-43d3-a6af-f826801ed969-proc") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
```

This shows that the deployment was picked up - and now the pods/containers etc are
starting up. This causes the load on the node to spike, and then:

```
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.472829    4382 kuberuntime_manager.go:404] No sandbox for pod "prop-prometheus-operator-operator-6bc6fd7c9d-d9qz4_default(47fba903-9c81-4c87-ab6c-9794e0829414)" can be found. Need to start a new one
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.491240    4382 reconciler.go:248] operationExecutor.MountVolume started for volume "sys" (UniqueName: "kubernetes.io/host-path/8861c2f4-e635-43d3-a6af-f826801ed969-sys") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.491289    4382 reconciler.go:248] operationExecutor.MountVolume started for volume "prop-prometheus-node-exporter-token-xkddz" (UniqueName: "kubernetes.io/secret/8861c2f4-e635-43d3-a6af-f826801ed969-prop-prometheus-node-exporter-token-xkddz") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.491318    4382 reconciler.go:248] operationExecutor.MountVolume started for volume "proc" (UniqueName: "kubernetes.io/host-path/8861c2f4-e635-43d3-a6af-f826801ed969-proc") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.491368    4382 operation_generator.go:713] MountVolume.SetUp succeeded for volume "proc" (UniqueName: "kubernetes.io/host-path/8861c2f4-e635-43d3-a6af-f826801ed969-proc") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.491567    4382 operation_generator.go:713] MountVolume.SetUp succeeded for volume "sys" (UniqueName: "kubernetes.io/host-path/8861c2f4-e635-43d3-a6af-f826801ed969-sys") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.499946    4382 operation_generator.go:713] MountVolume.SetUp succeeded for volume "prop-prometheus-node-exporter-token-xkddz" (UniqueName: "kubernetes.io/secret/8861c2f4-e635-43d3-a6af-f826801ed969-prop-prometheus-node-exporter-token-xkddz") pod "prop-prometheus-node-exporter-q7fvn" (UID: "8861c2f4-e635-43d3-a6af-f826801ed969")
Jan 30 23:27:56 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:27:56.555078    4382 kuberuntime_manager.go:404] No sandbox for pod "prop-prometheus-node-exporter-q7fvn_default(8861c2f4-e635-43d3-a6af-f826801ed969)" can be found. Need to start a new one
Jan 30 23:27:57 aks-agentpool-57418505-vmss000000 kubelet[4382]: 2020/01/30 23:27:57 [6925] [cni-ipam] Plugin azure-vnet-ipam version v1.0.29.
```

The deployment started at ~23:23 - the pods on this node (node 0 of a 5 node) are
really unhappy:

```
Jan 30 23:28:19 aks-agentpool-57418505-vmss000000 kubelet[4382]: 2020/01/30 23:28:19 [8711] [cni-ipam] Plugin stopped.
Jan 30 23:28:20 aks-agentpool-57418505-vmss000000 kubelet[4382]: W0130 23:28:20.210877    4382 pod_container_deletor.go:75] Container "2dcfc67a6a4774409a7200f91a13dd4fd37bd7d7723cc33207fe9817c981279e" not found in pod's containers
Jan 30 23:28:21 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:28:21.135911    4382 kubelet_pods.go:1090] Killing unwanted pod "prop-prometheus-operator-admission-patch-tbln5"
Jan 30 23:28:40 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:28:40.384030    4382 container_manager_linux.go:457] [ContainerManager]: Discovered runtime cgroups name: /system.slice/docker.service
Jan 30 23:33:15 aks-agentpool-57418505-vmss000000 kubelet[4382]: W0130 23:33:15.597465    4382 reflector.go:302] object-"kube-system"/"container-azm-ms-agentconfig": watch of *v1.ConfigMap ended with: too old resource version: 3245886 (3246372)
Jan 30 23:33:40 aks-agentpool-57418505-vmss000000 kubelet[4382]: I0130 23:33:40.384564    4382 container_manager_linux.go:457] [ContainerManager]: Discovered runtime cgroups name: /system.slice/docker.service
```
Here's the point of view from the container that had its sandbox fail:

```
jnoller@doge ~ $ -> (⎈ |kubernaughty:default)$ k logs prometheus-prop-prometheus-operator-prometheus-0 --all-containers --previous
level=info ts=2020-01-30T23:28:18.275Z caller=main.go:332 msg="Starting Prometheus" version="(version=2.13.1, branch=HEAD, revision=6f92ce56053866194ae5937012c1bec40f1dd1d9)"
level=info ts=2020-01-30T23:28:18.275Z caller=main.go:333 build_context="(go=go1.13.1, user=root@88e419aa1676, date=20191017-13:15:01)"
level=info ts=2020-01-30T23:28:18.275Z caller=main.go:334 host_details="(Linux 4.15.0-1064-azure #69-Ubuntu SMP Tue Nov 19 16:58:01 UTC 2019 x86_64 prometheus-prop-prometheus-operator-prometheus-0 (none))"
level=info ts=2020-01-30T23:28:18.275Z caller=main.go:335 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2020-01-30T23:28:18.275Z caller=main.go:336 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2020-01-30T23:28:18.277Z caller=main.go:657 msg="Starting TSDB ..."
level=info ts=2020-01-30T23:28:18.277Z caller=web.go:450 component=web msg="Start listening for connections" address=0.0.0.0:9090
level=info ts=2020-01-30T23:28:18.281Z caller=head.go:514 component=tsdb msg="replaying WAL, this may take awhile"
level=info ts=2020-01-30T23:28:18.282Z caller=head.go:562 component=tsdb msg="WAL segment loaded" segment=0 maxSegment=0
level=info ts=2020-01-30T23:28:18.283Z caller=main.go:672 fs_type=EXT4_SUPER_MAGIC
level=info ts=2020-01-30T23:28:18.283Z caller=main.go:673 msg="TSDB started"
level=info ts=2020-01-30T23:28:18.283Z caller=main.go:743 msg="Loading configuration file" filename=/etc/prometheus/config_out/prometheus.env.yaml
level=info ts=2020-01-30T23:28:18.283Z caller=main.go:526 msg="Stopping scrape discovery manager..."
level=info ts=2020-01-30T23:28:18.283Z caller=main.go:540 msg="Stopping notify discovery manager..."
level=info ts=2020-01-30T23:28:18.283Z caller=main.go:562 msg="Stopping scrape manager..."
level=info ts=2020-01-30T23:28:18.283Z caller=main.go:536 msg="Notify discovery manager stopped"
level=info ts=2020-01-30T23:28:18.283Z caller=main.go:522 msg="Scrape discovery manager stopped"
level=info ts=2020-01-30T23:28:18.283Z caller=manager.go:814 component="rule manager" msg="Stopping rule manager..."
level=info ts=2020-01-30T23:28:18.283Z caller=manager.go:820 component="rule manager" msg="Rule manager stopped"
level=info ts=2020-01-30T23:28:18.283Z caller=main.go:556 msg="Scrape manager stopped"
level=info ts=2020-01-30T23:28:18.294Z caller=notifier.go:602 component=notifier msg="Stopping notification manager..."
level=info ts=2020-01-30T23:28:18.294Z caller=main.go:727 msg="Notifier manager stopped"
level=error ts=2020-01-30T23:28:18.294Z caller=main.go:736 err="error loading config from \"/etc/prometheus/config_out/prometheus.env.yaml\": couldn't load configuration (--config.file=\"/etc/prometheus/config_out/prometheus.env.yaml\"): open /etc/prometheus/config_out/prometheus.env.yaml: no such file or directory"
Error from server (BadRequest): previous terminated container "prometheus-config-reloader" in pod "prometheus-prop-prometheus-operator-prometheus-0" not found
```

**Comment by jlogsdon** [comment]
I can't find this Disk IO report anywhere, is there documentation on enabling the report? We use terraform, for what its worth.

I can't enable public IP addresses to SSH into a box and review logs, but I can tell you that the tunnelfront container does not restart. I'm in the process of recreating a cluster, the steps to enable prometheus-operator some how hosed the last one I made.

**Comment by jnoller** [comment]
@jlogsdon We definitely need all of the log data from the nodes, but if you install the prometheus operator it should also show the key data. 

Just confirming, this only happens when running helm/tiller commands - if so that matches my tests, I can cause a cluster to lose nodes and API connectivity running `helm install istio` or `helm install prometheus-operator`, and that is triggered by the Disk IO spike on /dev/sda.

**Comment by jlogsdon** [comment]
The helm install command is how we've been replicating, yes. It's not an instant failure, the deployments  take 3-5 minutes to fully initialize after which some helm hooks would normally run. What we see instead is tunnelfront reconnect and the tiller "watch" appear to freeze and eventually the install times due to the wait. When this happens all of the deployments will finish starting, but the install will not continue. I've also seen it happen while waiting on a hook to complete.

When this happens to the tiller pod any `helm` commands (even a `list`) will fail until I restart the tiller pod.

### Logs

After getting SSH access sorted I ran through a test again. Frustratingly the issue did not occur, which is a first in my last 15 tests. I'm going to scale back the VMSS resources to what we initially had and see if it happens with that, and what the logs look like if it does.

**Comment by jnoller** [comment]
"The helm install command is how we've been replicating, yes. It's not an instant failure, the deployments take 3-5 minutes to fully initialize after which some helm hooks would normally run. What we see instead is tunnelfront reconnect and the tiller "watch" appear to freeze and eventually the install times due to the wait." 

What you described is exactly what I see - during the helm process, the docker image pull and startup swamp the IO for the root disk. Commands may freeze, timeout, crash, or get TCP/IP timeouts / dial errors.

**Comment by jlogsdon** [comment]
_sigh_ I ran our test again with D2s_v3 + 100GiB disk and the issue did not occur! I did see the load on the instance with tunnelfront reach ~7 or so, and some reports of slow `du` operations in kubelet logs.

The customer we are trying to resolve this issue with made an attempt about 4 hours ago (with a D4s_v3 + 1000GiB setup) and it failed with the timeout issue.

**Comment by jnoller** [comment]
@jlogsdon  Yeah - this will happen randomly - for example a working cluster will lose a node, that failure triggers IO spikes on the others which trips the throttle - then they die as well. What you may be seeing is a difference in logging, workload containers, etc - anything hitting the disk with IO (say, splunk or antivirus software). 

This is *10000000* dependent on workload + cluster configuration - both of your configurations/setups must match down to the host VM level.

**Comment by jlogsdon** [comment]
Thanks for all the help debugging this, we'd been at a loss for months so its nice to have some concrete information on whats going on.

How heavily has the `knode` approach been tested? We've already tried the first option, the second doesn't apply, and we've done what we can to reduce image size. I'm hesitant to recommend such a major change in how the cluster operates.

You say "our team" has developed the tool, but it's "owned" by an individual contributor. Is there any reason it's not under the Azure umbrella, given its specifically for resolving an issue with Azure's clusters?

**Comment by jnoller** [comment]
Knode is in use by myself and other large scale customers. That repo doesn't exist under AKS / Azure because of timing and it is not a part of the core product. We are looking at adding it this year.

You can offload the docker IO, or resize the clusters/nodepools to maximize to the 2tb disk limit to temporarily reduce the likelihood of this happening, or distribute the containers/workload across a wider number of nodes in the cluster

**Comment by jlogsdon** [comment]
@jnoller is the issue in https://github.com/Azure/AKS/issues/1373 resolved? I see its closed now, but it doesn't look like a resolution was found based on the final comments.

We're still encountering this issue with every new installation a customer runs.

**Comment by palma21** [comment]
AKS will be providing in the short term

a) Ephemeral OS (which much higher IOPS and a not the cost of a premium large SSD)
b) Docker Root configuration (same as knode functionality but embedded at start time)

The knode approach is a good workaround in the meant time but we're also happy to add you to the private preview of option b). If you're interested. Just open a Support Request and mention this issue.

Since both of the above options are captured on their own features in our roadmap and should solve this issue. 
https://github.com/Azure/AKS/issues/1370
https://github.com/Azure/AKS/issues/1753

Let me know if you feel it should stay open or have additional questions:

**Comment by 4c74356b41** [comment]
hey, I'm interested in both, I'm using knode already. when is the embedded knode release planned?
